File: include/PolyWorksSDK/samples/plugins/common/VisualBasic.NET/VBNetCommands/COM_SDK/Util/CComCommandParameterInfo.vb
' =================================================================================================
' Class describing a COM command parameter
'
' Copyright © InnovMetric Logiciels Inc. 2011 All Rights Reserved
' -------------------------------------------------------------------------------------------------
Namespace VBNetCommands

    ' =================================================================================
    ' COM command parameter class description implementation
    ' A description, a direction and a type must at least be specified
    ' Otherwise, only needed members require initialisation
    ' ---------------------------------------------------------------------------------
    Public Class CComCommandParameterInfo

        ' =============================================================================
        ' Constructor
        '
        ' Parameter: desc_ : Description
        ' Parameter: dir_  : Direction
        ' Parameter: type_ : Type
        ' -----------------------------------------------------------------------------
        Public Sub New(ByVal desc_ As String, _
                       ByVal dir_ As IMPluginLib.EParameterDirection, _
                       ByVal type_ As IMPluginLib.EParameterType)

            m_impl = New CComCommandParameterInfoImpl(desc_, dir_, type_)

        End Sub

        ' =============================================================================
        ' Copy constructor
        '
        ' Parameter: info_ : Object to be copied
        ' -----------------------------------------------------------------------------
        Public Sub New(ByVal info_ As CComCommandParameterInfo)

            If Not info_ Is Nothing Then

                m_impl = New CComCommandParameterInfoImpl(info_.DescGet(), info_.DirGet(), info_.TypeGet())

                ValuesSet(info_.ValuesGet())
                HasDoubleDefaultValueSet(info_.HasDoubleDefaultValueGet())
                DoubleDefaultValueSet(info_.DoubleDefaultValueGet())
                HasIntDefaultValueSet(info_.HasIntDefaultValueGet())
                IntDefaultValueSet(info_.IntDefaultValueGet())
                HasStringDefaultValueSet(info_.HasStringDefaultValueGet())
                StringDefaultValueSet(info_.StringDefaultValueGet())

            End If

        End Sub

        ' ==============================================================================
        ' Members access methods
        ' ------------------------------------------------------------------------------

        ' Description 
        Public Function DescGet() As String

            If Not m_impl Is Nothing Then

                Return m_impl.m_desc

            End If

            Return String.Empty

        End Function

        Public Sub DescSet(ByVal desc_ As String)

            If Not m_impl Is Nothing Then

                m_impl.m_desc = StrUtil.StrNonNull(desc_)

            End If

        End Sub

        ' Possible values (in a string, separated by comas)
        Public Function ValuesGet() As String

            If Not m_impl Is Nothing Then

                Return m_impl.m_values

            End If

            Return String.Empty

        End Function

        Public Sub ValuesSet(ByVal values_ As String)

            If Not m_impl Is Nothing Then

                m_impl.m_values = StrUtil.StrNonNull(values_)

            End If

        End Sub

        ' Direction
        Public Function DirGet() As IMPluginLib.EParameterDirection

            If Not m_impl Is Nothing Then

                Return m_impl.m_dir

            End If

            Return 0

        End Function

        Public Sub DirSet(ByVal dir_ As IMPluginLib.EParameterDirection)

            If Not m_impl Is Nothing Then

                m_impl.m_dir = dir_

            End If

        End Sub

        ' Type
        Public Function TypeGet() As IMPluginLib.EParameterType

            If Not m_impl Is Nothing Then

                Return m_impl.m_type

            End If

            Return 0

        End Function

        Public Sub TypeSet(ByVal type_ As IMPluginLib.EParameterType)

            If Not m_impl Is Nothing Then

                m_impl.m_type = type_

            End If

        End Sub

        ' True if parameter have a default double value
        Public Function HasDoubleDefaultValueGet() As Boolean

            If Not m_impl Is Nothing Then

                Return m_impl.m_hasDoubleDefaultValue

            End If

            Return False

        End Function

        Public Sub HasDoubleDefaultValueSet(ByVal hasValue_ As Boolean)

            If Not m_impl Is Nothing Then

                m_impl.m_hasDoubleDefaultValue = hasValue_

            End If

        End Sub

        ' Default value for double parameter
        Public Function DoubleDefaultValueGet() As Double

            If Not m_impl Is Nothing Then

                Return m_impl.m_doubleDefaultValue

            End If

            Return 0.0

        End Function

        Public Sub DoubleDefaultValueSet(ByVal value_ As Double)

            If Not m_impl Is Nothing Then

                m_impl.m_doubleDefaultValue = value_

            End If

        End Sub

        ' True if parameter have a default integer value
        Public Function HasIntDefaultValueGet() As Boolean

            If Not m_impl Is Nothing Then

                Return m_impl.m_hasIntDefaultValue

            End If

            Return False

        End Function

        Public Sub HasIntDefaultValueSet(ByVal hasValue_ As Boolean)

            If Not m_impl Is Nothing Then

                m_impl.m_hasIntDefaultValue = hasValue_

            End If

        End Sub

        ' Default value for integer parameter
        Public Function IntDefaultValueGet() As Integer

            If Not m_impl Is Nothing Then

                Return m_impl.m_intDefaultValue

            End If

            Return 0

        End Function

        Public Sub IntDefaultValueSet(ByVal value_ As Integer)

            If Not m_impl Is Nothing Then

                m_impl.m_intDefaultValue = value_

            End If

        End Sub

        ' True if parameter have a default string value
        Public Function HasStringDefaultValueGet() As Boolean

            If Not m_impl Is Nothing Then

                Return m_impl.m_hasStringDefaultValue

            End If

            Return False

        End Function

        Public Sub HasStringDefaultValueSet(ByVal hasValue_ As Boolean)

            If Not m_impl Is Nothing Then

                m_impl.m_hasStringDefaultValue = hasValue_

            End If

        End Sub

        ' Default value for string parameter
        Public Function StringDefaultValueGet() As String

            If Not m_impl Is Nothing Then

                Return m_impl.m_stringDefaultValue

            End If

            Return String.Empty

        End Function

        Public Sub StringDefaultValueSet(ByVal value_ As String)

            If Not m_impl Is Nothing Then

                m_impl.m_stringDefaultValue = StrUtil.StrNonNull(value_)

            End If

        End Sub

        Private m_impl As CComCommandParameterInfoImpl = Nothing

    End Class

End Namespace

File: include/PolyWorksSDK/samples/plugins/common/CSharp/CSharpCommands/COM_SDK/Util/CComCommandParameterInfo.cs
// =================================================================================================
// Class describing a COM command parameter
//
// Copyright © InnovMetric Logiciels Inc. 2009 All Rights Reserved
// -------------------------------------------------------------------------------------------------
namespace IM
{
    // =================================================================================
    // COM command parameter class description implementation
    // A description, a direction and a type must at least be specified
    // Otherwise, only needed members require initialisation
    // ---------------------------------------------------------------------------------
    public class CComCommandParameterInfo
    {
        // =============================================================================
        // Constructor
        //
        // Parameter: desc_ : Description
        // Parameter: dir_  : Direction
        // Parameter: type_ : Type
        // -----------------------------------------------------------------------------
        public CComCommandParameterInfo(string desc_, 
                                        IMPluginLib.EParameterDirection dir_, 
                                        IMPluginLib.EParameterType type_)
        {
            m_impl = new CComCommandParameterInfoImpl(desc_, dir_, type_);
        }

        // =============================================================================
        // Copy constructor
        //
        // Parameter: info_ : Object to be copied
        // -----------------------------------------------------------------------------
        public CComCommandParameterInfo(CComCommandParameterInfo info_)
        {
            if (info_ != null)
            {
                m_impl = new CComCommandParameterInfoImpl(info_.DescGet(), 
                                                          info_.DirGet(), 
                                                          info_.TypeGet());
                ValuesSet(info_.ValuesGet());
                HasDoubleDefaultValueSet(info_.HasDoubleDefaultValueGet());
                DoubleDefaultValueSet(info_.DoubleDefaultValueGet());
                HasIntDefaultValueSet(info_.HasIntDefaultValueGet());
                IntDefaultValueSet(info_.IntDefaultValueGet());
                HasStringDefaultValueSet(info_.HasStringDefaultValueGet());
                StringDefaultValueSet(info_.StringDefaultValueGet());
            }
        }

        // ==============================================================================
        // Members access methods
        // ------------------------------------------------------------------------------
        #region Member access methods

        // Description 
        public string DescGet()
        {
            if (m_impl != null)
            {
                return m_impl.m_desc;
            }
            return string.Empty;
        }
        public void DescSet(string desc_)
        {
            if (m_impl != null)
            {
                m_impl.m_desc = StrUtil.StrNonNull(desc_);
            }
        }

        // Possible values (in a string, separated by comas)
        public string ValuesGet()
        {
            if (m_impl != null)
            {
                return m_impl.m_values;
            }
            return string.Empty;
        }
        public void ValuesSet(string values_)
        {
            if (m_impl != null)
            {
                m_impl.m_values = StrUtil.StrNonNull(values_);
            }
        }

        // Direction
        public IMPluginLib.EParameterDirection DirGet()
        {
            if (m_impl != null)
            {
                return m_impl.m_dir;
            }
            return 0;
        }
        public void DirSet(IMPluginLib.EParameterDirection dir_)
        {
            if (m_impl != null)
            {
                m_impl.m_dir = dir_;
            }
        }

        // Type
        public IMPluginLib.EParameterType TypeGet()
        {
            if (m_impl != null)
            {
                return m_impl.m_type;
            }
            return 0;
        }
        public void TypeSet(IMPluginLib.EParameterType type_)
        {
            if (m_impl != null)
            {
                m_impl.m_type = type_;
            }
        }

        // True if parameter have a default double value
        public bool HasDoubleDefaultValueGet()
        {
            if (m_impl != null)
            {
                return m_impl.m_hasDoubleDefaultValue;
            }
            return false;
        }
        public void HasDoubleDefaultValueSet(bool hasValue_)
        {
            if (m_impl != null)
            {
                m_impl.m_hasDoubleDefaultValue = hasValue_;
            }
        }

        // Default value for double parameter
        public double DoubleDefaultValueGet()
        {
            if (m_impl != null)
            {
                return m_impl.m_doubleDefaultValue;
            }
            return 0.0;
        }
        public void DoubleDefaultValueSet(double value_)
        {
            if (m_impl != null)
            {
                m_impl.m_doubleDefaultValue = value_;
            }
        }

        // True if parameter have a default integer value
        public bool HasIntDefaultValueGet()
        {
            if (m_impl != null)
            {
                return m_impl.m_hasIntDefaultValue;
            }
            return false;
        }
        public void HasIntDefaultValueSet(bool hasValue_)
        {
            if (m_impl != null)
            {
                m_impl.m_hasIntDefaultValue = hasValue_;
            }
        }

        // Default value for integer parameter
        public int IntDefaultValueGet()
        {
            if (m_impl != null)
            {
                return m_impl.m_intDefaultValue;
            }
            return 0;
        }
        public void IntDefaultValueSet(int value_)
        {
            if (m_impl != null)
            {
                m_impl.m_intDefaultValue = value_;
            }
        }

        // True if parameter have a default string value
        public bool HasStringDefaultValueGet()
        {
            if (m_impl != null)
            {
                return m_impl.m_hasStringDefaultValue;
            }
            return false;
        }
        public void HasStringDefaultValueSet(bool hasValue_)
        {
            if (m_impl != null)
            {
                m_impl.m_hasStringDefaultValue = hasValue_;
            }
        }

        // Default value for string parameter
        public string StringDefaultValueGet()
        {
            if (m_impl != null)
            {
                return m_impl.m_stringDefaultValue;
            }
            return string.Empty;
        }
        public void StringDefaultValueSet(string value_)
        {
            if (m_impl != null)
            {
                m_impl.m_stringDefaultValue = StrUtil.StrNonNull(value_);
            }
        }

        #endregion

        private CComCommandParameterInfoImpl m_impl = null;
    }
}
